lib/search.rb

```diff
@@ -458,27 +458,13 @@ def self.advanced_filters
     end
   end
 
-  advanced_filter(/tags?:([a-zA-Z0-9,\-_+]+)/) do |posts, match|
-    if match.include?('+')
-      tags = match.split('+')
-
-      posts.where("topics.id IN (
-      SELECT tt.topic_id
-      FROM topic_tags tt, tags
-      WHERE tt.tag_id = tags.id
-      GROUP BY tt.topic_id
-      HAVING to_tsvector(#{default_ts_config}, array_to_string(array_agg(tags.name), ' ')) @@ to_tsquery(#{default_ts_config}, ?)
-      )", tags.join('&'))
-    else
-      tags = match.split(",")
+  advanced_filter(/^tags?:(.*)/) do |posts, match|
+    search_tags(posts, match)
+  end
 
-      posts.where("topics.id IN (
-      SELECT DISTINCT(tt.topic_id)
-      FROM topic_tags tt, tags
-      WHERE tt.tag_id = tags.id
-      AND tags.name in (?)
-      )", tags)
-    end
+  advanced_filter(/\-tags?:(.*)/) do |posts, match|
+    negative_tags = search_tags(posts, match)
+    posts.where(["posts.id NOT IN (?)", negative_tags.pluck("id")]).order("id")
```

Comment:
Hmm I'm quite uncomfortable with `negative_tags.pluck("id")` because it would trigger an extra SQL query. I think in `search_tags` we can just decide if it is a negative match or not and use `tags.name NOT IN` instead